Add the aix_inittab module to be able to modify the inittab at AIX systems

molekuul opened the pull request:
##### ISSUE TYPE
 - New Module Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
aix_inittab

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (devel 11ef87a034) last updated 2017/02/16 10:55:20 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

I created a new module to be able to add services to the inittab at AIX. With this module you can add, change and remove inittab entries at AIX with the AIX commands: mkitab, chitab and rmitab. If an entry allready exists, this is checked with the lsitab command, nothing will happen.

<!-- Paste verbatim command output below, e.g. before and after your change -->
```

```

Commit:
Added the aix_inittab module to be able to modify the inittab at AIX systems.

Commit:
fixed identation errors

Commit:
fixed identations and trailing whitespace
changed if conditions

Commit:
repaired tab-identation

Commit:
adjusted to long line and identation

Commit:
Adjusted the yaml in the example documentation

Commit:
Adjusted the documentation yaml

Commit:
Repaired return yaml

Commit:
repaired typo

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,261 @@
+# Import necessary libraries
+import itertools
+
+try:
```
Modules are not required to import json themselves.

molekuul replied:
I will remove this include.

molekuul replied:
I will remove the import of the json module, some old development stuff

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,261 @@
+                    (rc, out, err) = module.run_command(['mkitab', new_entry ])
+
+                if rc != 0:
+                    result['warnings']= "could not add"+" "+module.params['name']
```
Use module.warn() instead.

molekuul replied:
do you have an example how to use this?

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,261 @@
+        if current_entry['exist']:
+            (rc, out, err) = module.run_command(['rmitab',  module.params['name']])
+            if rc != 0:
+                result['warnings']= "could not remove"+" "+current_entry['name']
```
Use module.warn() instead.

dagwieers replied:
```
[dag@moria ansible.git]$ grep -r 'module.warn(' lib/ansible/modules/
lib/ansible/modules/system/systemd.py:        module.warn('The service (%s) is actually an init script but the system is managed by systemd' % unit)
```

molekuul replied:
thankx, I had to sync my fork.

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,261 @@
+                                  ], type='str'),
+            command = dict( required=True, type='str' ),
+            follow = dict( type='str' ),
+            action = dict( choices=[
```
This is the wrong way of doing Ansible. The user should describe the desired state, not instruct modules to perform actions. The module itself should determine what actions the specific state needs based on the current state and the desired state. That is what is needed for idempotency anyway.

So rather than action=install, it should be state=present.

Also, check-mode support and diff support are very important for a tool like this one IMO.

molekuul replied:
I will change this in the module: it will be state with choice: pressent or absent

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,261 @@
+    default: null
+  follow:
+    description: After which inittabline shoud the new entry follow.
+    required: False
```
required: False, and default: null can be left out. These are the default anyway, and it's easier to read when only the specific stuff is in the overview.

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,261 @@
+              ]
+    default: null
+notes:
+  - The changes are persisten across reboot.
```
typos: persistent / reboots

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,261 @@
+    required: True
+    type: string
+    default: null
+  processaction:
```
I think **processaction** is a poor name for this. I would stick to just **action** based on the documentation you pasted.

molekuul replied:
I will rename the proccessaction to action simular to the mkitab command, The action will be renamed to state as mentioned in comments below

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,261 @@
+    requred: True
+    type: string
+    default: null
+  follow:
```
Does the order really matter in an inittab ?

molekuul replied:
The inittab at AIX is run top down, it looks like:
[ Identifier ] : [ RunLevel ] : [ Action ] : [ Command ]
Depending on the action, the init proces will run a command once and not wait for it to finnish, or wil run a command once and wait for it. possible actions are:
* respawn:   If the process identified in this record does not exist, start the process. If the process currently exists, do nothing and continue scanning the /etc/inittab file.
* wait:  When the init command enters the run level specified for this record, start the process and wait for it to stop. While the init command is in the same run level, all subsequent reads of the /etc/inittab file ignore this object.
* once: When the init command enters the run level specified for this record, start the process, do not wait for it to stop and when it does stop do not restart the process. If the system enters a new run level while the process is running, the process is not restarted.
* boot: Read this record only when the system boots and reads the /etc/inittab file. The init command starts the process. Do not wait for the process to stop and when it does stop, do not restart the process. The run level for this process should be the default, or it must match the run level specified by the init command at startup time.
* bootwait: Read this record only when the system boots and reads the /etc/inittab file. The init command starts the process. Wait for it to stop, and when it does stop, do not restart the process.
* powerfail: Start the process identified in this record only when the init command receives a SIGPWR power fail signal.
* powerwait: Start the process identified in this record only when the init command receives a SIGPWR power fail signal, and wait until it stops before continuing to process the /etc/inittab file.
* off: If the process identified in this record is currently running, send the warning signal SIGTERM and wait 20 seconds before sending the SIGKILL kill signal. If the process is nonexistent, ignore this line.
* hold: When the process identified in this record is terminated, do not start a new one. The hold action can only be activated by the phold command.
* ondemand: Functionally identical to respawn. If the process identified in this record does not exist, start the process. If the process currently exists, do nothing and continue scanning the /etc/inittab file. Specify this action to perform the respawn action when using a, b, or c run levels.
* initdefault: A line with this action is processed only when the init command is originally invoked. The init command uses this line to determine which run level to originally enter. The command does this by taking the highest run level specified in the RunLevel parameter and using that as the command's initial state. If the RunLevel parameter is empty, its value is interpreted as 0123456789, and the init command enters a run level of 9. If the init command does not find an initdefault line in the inittab file, it requests an initial run level from the operator at initial program load (IPL) time.
* sysinit: Start the process identified in this record before the init command tries to access the console. For example, you might use this to initialize devices.

dagwieers replied:
That the implementation runs top-down is without a doubt, but it doesn't really process the items sequentially. All items for the same runlevel are started in parallel. So the order should not matter and therefore there is no need to ask the user IMO. So I would remove that option.

dagwieers replied:
BTW The most important reason for not adding that parameter, is because it can cause real idempotency issues.

molekuul replied:
The run is actually sequential, especially when there is a wait involved, it runs the command, and waits until it is finished before continuing reading the inittab file. if there are dependancies between services to start you have to place the entry at the right place in the inittab.

How do you think this follow parameter can cause idempotency issues?

dagwieers replied:
Alright, so if the order is important there is no way around this.
The following playbook is problematic:
```yaml
- name: Ensure service1 exists
  aix_inittab:
    name: service1
    runlevel: 4
    action: once
    command: echo service1
    insertafter: foobar
    state: present
- name: Ensure service2 exists
  aix_inittab:
    name: service2
    runlevel: 3
    action: once
    command: echo service2
    insertafter: foobar
    state: present
```
So both actions will always cause a change when re-run.
BTW what happens if foobar does not exist ?

dagwieers replied:
Please rename **follow** to **insertafter**, that is an existing parameter already used by **lineinfile** and better describes what it does.

molekuul replied:
In the mentioned playbook, service1 will be inserted after foobar, and then service2 will be inserted after foobar, so the order will be:
foobar
service2
service1

When the playbook is ran again, nothing will happen, the follow parameter is not included in the changing of inittab entries. If the ident exists, only the runlevel, processaction or command will be changed, if one of them is different.
see code:
```
or (  
 219   +                module.params['runlevel'] != current_entry['runlevel'] or  
 220   +                module.params['processaction'] != current_entry['processaction'] or  
 221   +                module.params['command'] != current_entry['command'] ): 
```

If foobar does not exists, the returncode of the mkitab command is 1 and an error is generated.

Commit:
Removed unnecessary parameters from comment
changed action to state, and changed the way to call this module
changed proccessaction to action

Commit:
adjusted result['warnings'] to module.warn()

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,240 @@
+    type: string
+  state:
+    description: Whether the entry should be present or absent in the inittab file
+    required: True
```
Parameter `state` shouldn't be required as it defaults to `present`.

I also noticed that you require name, runlevel, action and command, also in the case of `state=absent`. That might as well be intentional, but often removing stuff only requires the bare minimum for matching the entry (in this case maybe command and/or action is not required to remove the entry ?).

molekuul replied:
I adjusted ```state: required ```  in the code, but i forgot to adjust the documentation

You are right, when state is absent the runlevel, action and command are not required. I could not find how to program this.

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,240 @@
+  - The changes are persistent across reboots.
+  - You need root rights to read or adjust the inittab with the lsitab, chitab,
+    mkitab or rmitab commands.
+  - tested at AIX 7.1.
```
Should be:
- Tested on AIX v7.1

Commit:
adjusted the documentation

molekuul commented:
ready_for_review

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,239 @@
+    type: string
+  command:
+    description: What command has to run.
+    requred: True
```
Typo: required

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,239 @@
+                if rc != 0:
+                    module.warn('could not add (%s)' % current_entry['name'])
+                result['status'] = "add inittab entry"+" "+module.params['name']
+                result['changed'] = True
```
Looking at the logic, I am confused by why we are only sending out a warning when an entry cannot be added.
The correct action here is to fail.

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,239 @@
+            if current_entry['exist']:
+                (rc, out, err) = module.run_command(['chitab', new_entry])
+                if rc != 0:
+                    module.warn('could not change (%s)' % current_entry['name'])
```
Looking at the logic, I am confused by why we are only sending out a warning when an entry cannot be added.
The correct action here is to fail.

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,239 @@
+        if current_entry['exist']:
+            (rc, out, err) = module.run_command(['rmitab',  module.params['name']])
+            if rc != 0:
+                module.warn('could not remove (%s)' % current_entry['name'])
```
Looking at the logic, I am confused by why we are only sending out a warning when an entry cannot be added.
The correct action here is to fail.

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,239 @@
+        'name':  module.params['name'],
+        'changed': False,
+        'status': {},
+        'msg': [],
```
What is the purpose of `msg` here, it does not seem to be used ?

dagwieers commented on lib/ansible/modules/system/aix_inittab.py:
```diff
@@ -0,0 +1,239 @@
+    result = {
+        'name':  module.params['name'],
+        'changed': False,
+        'status': {},
```
It seems that `status` is used as a normal string, but it was defined here as a dictionary ?

Also, you don't have to initialize these entries here, you can simply assign them at will. Unless you have cases where you prefer to return an empty string rather than no entry.

dagwieers commented:
A few other suggestions/ideas:

- You can implement check-mode support fairly easy (by wrapping each command with a conditional statement). It will not correctly identify errors when running the command, but that's only possible if the command itself has some sort of dry-run mode. Often Unix commands do not have it.
- It would be nice to have diff support in the module (so you can test what would happen to the file and see the diff output to verify what you're doing is correct). Adding diff support for this module is fairly simple when not run in check-mode, however diff support *must* support check-mode when implemented. So the only way you can support diff-mode properly is to work on a copy of the inittab rather than working on the inittab itself, and then report the difference before and after. I don't know if this is even possible. I guess that's the drawback of using commands rather than interact with the file directly. (And I am not saying that would be a better approach overal BTW)

tmshn commented:
Another suggestion: could you consider using an auto-formatter? It makes your code more conventional and readable for other collaborators.

One choice is: https://github.com/hhatto/autopep8
(Of course, you can choose as you like. Some tools are integrated to an editor like Sublime Text, Atom, Emacs or PyCharm...etc; that's quite useful.)

molekuul commented:
At AIX, the inittab must not be adjusted by changing the inittab file, the init proces has the inittab open in memory, I have encountered issues where one edited the inittab file, but ended-up with an empty inittab file when a reboot was issued. 
IBM states you have to use the mkitab, rmitab and chitab commands when changing the inittab.

dagwieers commented:
@molekuul That's what I assumed if they had these specialized commands for it.

Commit:
changed warning to failed
ran module against autopep8

Commit:
added check_mode

Commit:
fixed typo

Commit:
changed description and short description, to be able to push again after a failure at shippable

molekuul commented:
ready_for_review

dagwieers commented:
shipit

molekuul commented:
Hi,

As this is my first real contribution to ansible what should I do now to get this module merged in ansible?

Kind regards,

Joris

dagwieers commented:
@molekuul Congratz ! :+1:

molekuul commented:
Thankx !
